src/services/file.c

```diff
@@ -116,15 +116,16 @@ int nspire_file_read(nspire_handle_t *handle, const char *path,
 		if ( (ret = data_read(handle, buffer, len+1, &len)) )
 			goto end;
 
-		if (size) {
-			size_t to_copy = len - 1;
-			memcpy(ptr, buffer + 1, to_copy);
-			size -= to_copy;
-
-			ptr += to_copy;
+		size_t to_copy = len - 1;
+		memcpy(ptr, buffer + 1, (size < to_copy) ? size : to_copy);
+		if (total_bytes) *total_bytes += (size < to_copy) ? size : to_copy;
+		size -= (size < to_copy) ? size : to_copy;
+		if(!size) {
+			goto end;
```

Comment:
This means that when an insufficient size was passed, it aborts early, which might be a protocol error. This also repeats the same code three times in a row.

Reply:
It *should* abort early—if the file is lengthened on the calculator between getting its attributes and reading it, this will just write past the buffer, i.e. UB.

Reply:
It shouldn't keep writing past the buffer end, but keep reading data from the calc.
That's what the previous version did, modulo integer overflow if `size < to_copy`.

Reply:
I'm not opposed to change the behaviour here if testing proves that it's ok, but IMO it shouldn't be in this PR